File: apps/api/src/services/outreachHook.ts
import type { CompanyIntel, HookConfidence } from "@believe-ai/shared";

/**
 * Step A — grounded hook selection. Deliberately makes no LLM call: every
 * hook here traces to a real, already-known fact — a skill the job requires
 * that the company's own tech stack also uses, or a high-confidence company
 * milestone. Ported from HireConnect's personalization_logic.py, adapted:
 * warm-path hooks (from a discovered contact's shared alma mater / prior
 * company) are deferred to Phase 3 (lead_discovery) since no contact
 * discovery exists yet — every hook in this phase is job/company-grounded,
 * not contact-specific, and so is currently the same across contacts in one
 * generation batch. That's an honest consequence of not having per-contact
 * signals yet, not a bug.
 */
export interface OutreachHook {
  hook: string;
  confidence: HookConfidence;
}

export function computeOutreachHook(company: string, jobSkills: string[], companyIntel: CompanyIntel): OutreachHook {
  const jobSkillsLower = new Set(jobSkills.map((s) => s.toLowerCase()));
  const techOverlap = companyIntel.techStack.find((t) => jobSkillsLower.has(t.toLowerCase()));
  if (techOverlap) {
    return {
      hook: `Shared tech stack: this role's use of ${techOverlap} lines up with ${company}'s stack.`,
      confidence: "high",
    };
  }

  if (companyIntel.funding && companyIntel.confidence.funding === "high") {
    return { hook: `Recent milestone: ${company}'s ${companyIntel.funding} funding round.`, confidence: "high" };
  }

  if (companyIntel.hiringTrend && companyIntel.confidence.hiringTrend === "high") {
    return { hook: `Company signal: ${company} is currently ${companyIntel.hiringTrend}.`, confidence: "high" };
  }

  return { hook: `No distinguishing public fact found for ${company} yet.`, confidence: "low" };
}

/** Skills the job requires that the candidate's own resume actually mentions — never the gaps. */
export function computeMatchingSkills(jobSkills: string[], resumeContent: string | null): string[] {
  if (!resumeContent) return [];
  const lowered = resumeContent.toLowerCase();
  return jobSkills.filter((skill) => lowered.includes(skill.toLowerCase()));
}

File: apps/backend/services/outreach_hook.py
"""Step A — grounded hook selection. Deliberately makes no LLM call: every
hook here traces to a real, already-known fact — a skill the job requires
that the company's own tech stack also uses, or a high-confidence company
milestone. Mirrors apps/api's services/outreachHook.ts exactly."""

from __future__ import annotations

from dataclasses import dataclass

from models.outreach_draft import HookConfidence
from schemas.ai import CompanyIntelResult


@dataclass
class OutreachHook:
    hook: str
    confidence: HookConfidence


def compute_outreach_hook(company: str, job_skills: list[str], company_intel: CompanyIntelResult) -> OutreachHook:
    job_skills_lower = {s.lower() for s in job_skills}
    tech_overlap = next((t for t in company_intel.techStack if t.lower() in job_skills_lower), None)
    if tech_overlap:
        return OutreachHook(
            hook=f"Shared tech stack: this role's use of {tech_overlap} lines up with {company}'s stack.",
            confidence="high",
        )

    if company_intel.funding and company_intel.confidence.funding == "high":
        return OutreachHook(hook=f"Recent milestone: {company}'s {company_intel.funding} funding round.", confidence="high")

    if company_intel.hiringTrend and company_intel.confidence.hiringTrend == "high":
        return OutreachHook(hook=f"Company signal: {company} is currently {company_intel.hiringTrend}.", confidence="high")

    return OutreachHook(hook=f"No distinguishing public fact found for {company} yet.", confidence="low")


def compute_matching_skills(job_skills: list[str], resume_content: str | None) -> list[str]:
    """Skills the job requires that the candidate's own resume actually mentions — never the gaps."""
    if not resume_content:
        return []
    lowered = resume_content.lower()
    return [skill for skill in job_skills if skill.lower() in lowered]
